File: server/app/controllers/comments_controller.py
from flask import jsonify, g
from app import db
from app.models import Post, Comment, User


def get_comments(post_title):
    post = Post.query.filter_by(title=post_title).first()
    if not post:
        return jsonify({"error": "Post not found."}), 404

    comments = (
        Comment.query.filter_by(post_id=post.id)
        .join(User)
        .add_columns(User.username, Comment.id, Comment.text, Comment.created_at)
        .order_by(Comment.created_at.desc())
        .all()
    )

    result = []
    for comment in comments:
        result.append(
            {
                "id": comment.id,
                "text": comment.text,
                "username": comment.username,
                "createdAt": comment.created_at,
            }
        )

    return jsonify(result), 200


def create_comment(post_title, data):
    post = Post.query.filter_by(title=post_title).first()
    if not post:
        return jsonify({"error": "Post not found."}), 404

    new_comment = Comment(
        text=data.get("text"),
        user_id=g.current_user.id,
        post_id=post.id,
    )
    db.session.add(new_comment)
    db.session.commit()

    return jsonify({"message": "Comment created successfully."}), 201


def edit_comment(post_title, comment_id, data):
    comment = Comment.query.filter_by(id=comment_id).first()
    if not comment:
        return jsonify({"error": "Comment not found."}), 404

    if comment.user_id != g.current_user.id:
        return jsonify({"error": "You are not authorized to edit this comment."}), 403

    comment.text = data.get("text")
    db.session.commit()

    return jsonify({"message": "Comment successfully edited."}), 201


def delete_comment(post_title, comment_id):
    comment = Comment.query.filter_by(id=comment_id).first()
    if not comment:
        return jsonify({"error": "Comment not found."}), 404

    if comment.user_id != g.current_user.id:
        return jsonify({"error": "You are not authorized to delete this comment."}), 403

    db.session.delete(comment)
    db.session.commit()

    return jsonify({"message": "Comment deleted successfully."}), 200

File: server_node/controllers/commentsController.js
const { PrismaClient } = require("@prisma/client");
const prisma = new PrismaClient();
const asyncHandler = require("express-async-handler");
const CustomError = require("../utils/customError");
const { handleValidationErrors } = require("../utils/validator");

const getComments = asyncHandler(async (req, res) => {
  const postSlug = decodeURIComponent(req.params.postSlug);
  const post = await prisma.post.findUnique({ where: { slug: postSlug } });

  if (!post) {
    throw new CustomError(404, "Post not found.");
  }

  const comments = await prisma.comment.findMany({
    where: { postId: post.id },
    include: {
      user: {
        select: { username: true },
      },
    },
    orderBy: { createdAt: "desc" },
  });

  comments.forEach((comment) => {
    comment.username = comment.user.username;
    delete comment.user;
  });

  res.status(200).json(comments);
});

const createComment = asyncHandler(async (req, res) => {
  handleValidationErrors(req);

  const postSlug = decodeURIComponent(req.params.postSlug);
  const post = await prisma.post.findUnique({ where: { slug: postSlug } });

  if (!post) {
    throw new CustomError(404, "Post not found.");
  }

  await prisma.comment.create({
    data: {
      text: req.body.text,
      userId: req.user.id,
      postId: post.id,
      updatedAt: new Date(),
    },
  });

  res.status(201).json({ message: "Comment created successfully." });
});

const editComment = asyncHandler(async (req, res) => {
  handleValidationErrors(req);

  const commentId = parseInt(req.params.commentId, 10);
  if (isNaN(commentId)) {
    throw new CustomError(400, "Invalid comment ID.");
  }

  const comment = await prisma.comment.findUnique({
    where: { id: commentId },
  });

  if (!comment) {
    throw new CustomError(404, "Comment not found.");
  }

  if (comment.userId !== req.user.id) {
    throw new CustomError(403, "You are not authorized to edit this comment.");
  }

  await prisma.comment.update({
    where: { id: commentId },
    data: {
      text: req.body.text,
      updatedAt: new Date(),
    },
  });

  res.status(201).json({ message: "Comment successfully edited." });
});

const deleteComment = asyncHandler(async (req, res) => {
  const commentId = parseInt(req.params.commentId, 10);
  if (isNaN(commentId)) {
    throw new CustomError(400, "Invalid comment ID.");
  }

  const comment = await prisma.comment.findUnique({
    where: { id: commentId },
  });

  if (!comment) {
    throw new CustomError(404, "Comment not found.");
  }

  if (comment.userId !== req.user.id) {
    throw new CustomError(403, "You are not authorized to delete this comment.");
  }

  await prisma.comment.delete({
    where: { id: commentId },
  });

  res.status(200).json({ message: "Comment deleted successfully." });
});

module.exports = {
  getComments,
  createComment,
  editComment,
  deleteComment,
};
